### ISSUE TITLE
Error in fusing layers

### ISSUE THREAD

**Issue by miziworld** [espnet/espnet#4147]
Hello, I am trying to implement PTQ(Post training quantization).
Among them, layer fusion is essential to proceed with static quantization. 
When using the E2E conformer model of espnet1, conv, linear, and relu cannot be fused.
The code and attribute error details are attached. Please advise on how to solve it.

in espnet/pytorch_backend/recog.py
![image](https://user-images.githubusercontent.com/17849068/157043336-e2b8ea2a-281d-47f2-8dfb-bde9aabaff2b.png)

in error
![image](https://user-images.githubusercontent.com/17849068/157043446-9bef8ed9-c678-4f11-8c31-944f612ce7d2.png)

**Comment by b-flo** [comment]
Hi,

Looking at the [tutorial for fuse modules](https://pytorch.org/tutorials/recipes/fuse.html) and docs, the second parameter for `fuse_modules` is a "list of list of module names (to fuse)". Also, there are no modules named "relu" or "linear" in `model` (i.e.: `model.relu`), you will need to specify the correct modules (e.g.: `model.enc.enc.nbrnn` with LSTM encoder)

**Comment by miziworld** [comment]
Thank you for answer.
I'm using a conformer model and I want to fuse conv and relu or conv and linear modules in an encoder.
Could you please help me with a little more detail on where I am going to add the code?

**Comment by b-flo** [comment]
> Could you please help me with a little more detail on where I am going to add the code?

Hum, I'm not really sure what's the better approach for static quantization in ESPnet yet.

Also, I can't comment on the whole procedure nor the specifics but for `fuse_modules`, it seems you just need to load the model and do the module fusing:

```
model, train_args = load_trained_model(args.model)
model_fused = torch.quantization.fuse_modules(model, [['module_name1', 'module_name2', ...]], inplace=False)
```

If you're not sure about the names, you can just print the model to get them.

**Comment by miziworld** [comment]
thank you.
In general, when using the code you suggested, there should be no problem, but attribute errors keep occurring, so I referred to the PyTorch tutorial.
(https://pytorch.org/docs/stable/_modules/torch/ao/quantization/fuse_modules.html#fuse_modules)
There are steps to proceed with fusion compared to before, but I am not sure how to solve the problem of 'E2E' object has no attribute 'torch' error. Can you help?

this is modified code:
![image](https://user-images.githubusercontent.com/17849068/157324773-70f3357f-a589-4137-a4fe-6640e0474dfe.png)
and this is the error message
![image](https://user-images.githubusercontent.com/17849068/157325046-2e822131-4683-49b6-a695-56bec3fe410d.png)

**Comment by miziworld** [comment]
Thank you so much for your kind and detailed explanation. 
Thanks for your help, I succesfully fused modules  i want like this.
![image](https://user-images.githubusercontent.com/17849068/157657943-5b723551-d302-4be5-8bc7-95c26f7418fe.png)
I fused modules ['model.encoder.embed.conv.0', 'model.encoder.embed.conv.1'] in conformer encoder.

But when I run it as is, it shows the following error:
![image](https://user-images.githubusercontent.com/17849068/157657720-3d3c2687-3d31-42a7-9eb6-23a4e779501a.png)


Searching for cases with similar errors
1. It is said that the QuantStub() and DeQuantStub() layers should be included when defining the model. How can I add this?
(https://spell.ml/blog/pytorch-quantization-X8e7wBAAACIAHPhT)
2. The conv module seems to use quint8 as the quantization config for activation such as int8 and relu. How should I modify it like the following link?
(https://github.com/pytorch/pytorch/blob/1edf6f56477ea317af845a4cd65eb311737961f0/torch/ao/quantization/qconfig.py#L141)
Your help is really helping me a lot. Thank you for always.
4.

**Comment by b-flo** [comment]
> model_prepared=torch.quantizaion.prepare(model_fused) # when I delete this line, module fusion is successfully done

Fusion is successful if `fuse_modules` is successful, `prepare` is unrelated to that, no? 

> Removing the model_fused line causes fusion but does not reduce the number or size of model parameters when printing the model. Is this normal?

I'm not sure I understand, are you asking why fusion doesn't reduce number of parameters or model size?
Correct me if I'm wrong but from my (early) understanding, fusing is mainly used to improve memory usage/access and numerical accuracy for the fused operations.

> Searching for cases with similar errors

Also, I'm sorry but as I said, I can't comment on the specifics of static quantization because I'm not up-to-date with the documentation nor the correct procedure in regards to ESPnet1 or ESPnet2 models. I'll let others answer from there.

**Comment by tarun73** [comment]
@miziworld where you able to solve your issue.
i am facing a similar issue with conformer 

- embedding  layers require weight observer with dtype=torch.quint8
- convolution layers require weight observer with dtype=torch.qint8

unable to find a way to send different dtypes for different layers.
